### PR TITLE
[Merged by Bors] - feat: port NumberTheory.ADEInequality

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mo271** [comment]
Remaining open questions

- [ ]  Is it ok to use uppercase for the letters "A" "D" and "E" in defs and theorems, like I propose?
- [ ] Some proofs need a bit of golf, most notably and similar places. It works for now, but the more elegant version in lean3 didn't work, perhaps mainly because of the missing `coe_coe`
  https://github.com/leanprover-community/mathlib4/blob/018c40bcbe97f0ab5e43be037415259e34ab3cd1/Mathlib/NumberTheory/ADEInequality.lean#L165-L168
- [ ] One proof does not work that uses `interval_cases` extensively. This is because of this  commented out line:
  https://github.com/leanprover-community/mathlib4/blob/c16c0c414d4ea87bd10716ffd38b93d9d66aa215/Mathlib/Tactic/IntervalCases.lean#L287
  We probably want to fix that tactic?

File: Mathlib/NumberTheory/ADEInequality.lean [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  rintro (⟨p', q', H⟩ | ⟨n, H⟩ | H | H | H)
+  · rw [← H, A', sumInv_pqr, add_assoc]
+    simp only [lt_add_iff_pos_right, PNat.one_coe, inv_one, Nat.cast_one]
+    apply add_pos <;> simp only [PNat.pos, Nat.cast_pos, inv_pos]
+  · rw [← H, D', sumInv_pqr]
+    conv_rhs => simp only [OfNat.ofNat, PNat.mk_coe]
+    norm_num
+  all_goals
+    rw [← H, E', sumInv_pqr]
+    conv_rhs => simp only [OfNat.ofNat, PNat.mk_coe]
```
Comment (Komyyy) [inline Mathlib/NumberTheory/ADEInequality.lean:174]: We can golf here by `conv_rhs => simp only [OfNat.ofNat, PNat.mk_coe]`.

Comment (mo271) [reply Mathlib/NumberTheory/ADEInequality.lean:174]: great!

File: Mathlib/NumberTheory/ADEInequality.lean [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+  interval_cases p
+  · exact admissible_A' q r
+  have hq4 : q < 4 := lt_four hqr H
+  interval_cases
+  · exact admissible_D' r
+  have hr6 : r < 6 := lt_six H
+  interval_cases
+  · exact admissible_E6
+  · exact admissible_E7
+  · exact admissible_E8
```
Comment (Komyyy) [inline Mathlib/NumberTheory/ADEInequality.lean:248]: If `norm_num` should fail to support `ℕ+`, try this.
```suggestion
theorem admissible_of_one_lt_sumInv_aux' {p q r : ℕ+} (hpq : p ≤ q) (hqr : q ≤ r)
    (H : 1 < sumInv {p, q, r}) : Admissible {p, q, r} := by
  have hp3 : p < 3 := lt_three hpq hqr H
  -- Porting note: `interval_cases` doesn't support `ℕ+` yet.
  replace hp3 := Finset.mem_Iio.mpr hp3
  conv at hp3 => change p ∈ ({1, 2} : Multiset ℕ+)
  fin_cases hp3
  · exact admissible_A' q r
  have hq4 : q < 4 := lt_four hqr H
  replace hq4 := Finset.mem_Ico.mpr ⟨hpq, hq4⟩; clear hpq
  conv at hq4 => change q ∈ ({2, 3} : Multiset ℕ+)
  fin_cases hq4
  · exact admissible_D' r
  have hr6 : r < 6 := lt_six H
  replace hr6 := Finset.mem_Ico.mpr ⟨hqr, hr6⟩; clear hqr
  conv at hr6 => change r ∈ ({3, 4, 5} : Multiset ℕ+)
  fin_cases hr6
  · exact admissible_E6
  · exact admissible_E7
  · exact admissible_E8
```

Comment (Komyyy) [reply Mathlib/NumberTheory/ADEInequality.lean]: It seems that handling `ℕ+` by `norm_num` is difficult, so I commited this.

Comment (mo271) [reply Mathlib/NumberTheory/ADEInequality.lean]: great! No need for interval_cases anymore

Comment (mo271) [reply Mathlib/NumberTheory/ADEInequality.lean]: Is this ready for review then, @Komyyy?

**Comment by Komyyy** [comment]
I think it is OK to use uppercase because `Combinator.S`, `K`, `I` are in `Mathlib.Init.Core`.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**